core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -540,4 +556,100 @@ class EntityService(protected val ctx: RawlsRequestContext,
+                                                              SamWorkspaceActions.own,
+                                                              Some(WorkspaceAttributeSpecs(all = false))
+      )
+
+      // confirm if this is already a quicksilver workspace by checking settings
+      workspaceSettingService = workspaceSettingServiceConstructor.get.apply(ctx)
+      settings <- workspaceSettingService.getWorkspaceSettings(workspaceName)
+      _ = if (
+        settings
+          .find(_.isInstanceOf[CompactDataTablesSetting])
+          .asInstanceOf[Option[CompactDataTablesSetting]]
+          .exists(_.config.enabled)
+      ) {
+        throw new RawlsExceptionWithErrorReport(ErrorReport("Quicksilver already enabled for this workspace"))
+      }
+
+      // get the local (legacy) provider
+      localProvider <- entityManager.resolveProviderFuture(EntityRequestArguments(workspaceContext, ctx))
+
+      // change the workspace to be quicksilver-enabled
+      _ <- workspaceSettingService.setWorkspaceSettings(
+        workspaceName,
+        List(CompactDataTablesSetting(CompactDataTablesConfig(enabled = true)))
+      )
+
+      // get the list of entity types in this workspace
+      entityTypeMetadata <- localProvider.entityTypeMetadata(useCache = true, ctx)
+
+      // start a transaction; here's where we do a bunch of writes
+      _ <- dataSource.inTransaction { dataAccess =>
+        val shardId: String = dataAccess.determineShard(workspaceContext.workspaceIdAsUUID)
+
+        // loop over entity types
+        def allTypesResult: Iterable[ReadWriteAction[Seq[Int]]] =
+          entityTypeMetadata.map { case (entityType, metadata) =>
+            logger.info(s"Quicksilver migration:     - $entityType (${metadata.count}) ...")
+
+            val thisTypeList: ReadAction[Seq[Entity]] = DBIO.from(
+              localProvider
+                .listEntities(entityType)
+                .runWith(Sink.seq)
+            )
+
+            val thisTypeInserts: ReadWriteAction[Seq[Int]] = thisTypeList flatMap { entities =>
+              DBIO.sequence(entities.map { entity =>
+                // ... insert into the temp table
+                // TODO CORE-364: batch these inserts?
+                dataAccess.compactEntityQuery.migrationInsertAttributesToTempTable(entity)
+              })
+            }
+
+            thisTypeInserts
+          }
+
+        for {
+          // create temp table
+          _ <- dataAccess.compactEntityQuery.migrationCreateTempTable
+          // insert entity name, entity type, and attributes to the temp table
+          _ = logger.info(s"Quicksilver migration: inserting to temp table ...")
+          _ <- DBIO.sequence(allTypesResult)
+          // update ENTITY from the contents of the temp table
+          _ = logger.info(s"Quicksilver migration: updating ENTITY from temp table ...")
+          _ <- dataAccess.compactEntityQuery.migrationUpdateFromTempTable(workspaceContext.workspaceIdAsUUID)
```

Comment:
the pattern of inserting into a temp table and then updating `ENTITY` based on that temp table allows us to later optimize by batching up all the inserts. The alternative is to issue one `update ENTITY` statement for each entity being migrated; `update` statements cannot be batched.